distributed/s3.py

```diff
@@ -26,148 +32,525 @@
+        self.bucket = bucket
+        self.key = key
+        self.blocksize = block_size
+        self.cache = b""
+        self.loc = 0
+        self.start = None
+        self.end = None
+        self.closed = False
+
+    def info(self):
+        return self.s3.info(self.path)
+
+    def tell(self):
+        return self.loc
+
+    def seek(self, loc, whence=0):
+        if whence == 0:
+            self.loc = loc
+        elif whence == 1:
+            self.loc += loc
+        elif whence == 2:
+            self.loc = self.size + loc
+        else:
+            raise ValueError("invalid whence (%s, should be 0, 1 or 2)" % whence)
+        if self.loc < 0:
+            self.loc = 0
+        return self.loc
+
+    def _fetch(self, start, end):
         try:
-            return get_s3(True)
-        except:
-            return get_s3(False)
-
-    with get_s3_lock:
-        key = anon, get_thread_identity()
-        if not _conn.get(key):
-            logger.debug("Open S3 connection.  Anonymous: %s.  Thread ID: %d",
-                         *key)
-            s3 = boto3.resource('s3')
-            if anon:
-                s3.meta.client.meta.events.register('choose-signer.s3.*',
-                        disable_signing)
-            _conn[key] = s3
-        return _conn[key]
-
-
-def get_list_of_summary_objects(bucket_name, prefix='', delimiter='',
-        page_size=DEFAULT_PAGE_LENGTH, anon=None):
-    s3 = get_s3(anon)
-    if bucket_name.startswith('s3://'):
-        bucket_name = bucket_name[len('s3://'):]
-    if prefix.startswith('/'):
-        prefix = prefix[1:]
-
-    L = list(s3.Bucket(bucket_name)
-               .objects.filter(Prefix=prefix, Delimiter=delimiter)
-               .page_size(page_size))
-    return [s for s in L if s.key[-1] != '/']
-
-
-def read_content_from_keys(bucket, key, anon=None):
-    if bucket.startswith('s3://'):
-        bucket = bucket[len('s3://'):]
-    s3 = get_s3(anon)
-    return s3.Object(bucket, key).get()['Body'].read()
-
-
-def read_bytes(bucket_name, prefix='', path_delimiter='', executor=None,
-               lazy=True, anon=None):
-    """ Read data on S3 into bytes in distributed memory
+            if self.start is None and self.end is None:
+                # First read
+                self.start = start
+                self.end = end + self.blocksize
+                self.cache = self.s3.s3.get_object(Bucket=self.bucket, Key=self.key,
+                                                Range='bytes=%i-%i' % (start, self.end - 1)
+                                                )['Body'].read()
+            if start < self.start:
+                new = self.s3.s3.get_object(Bucket=self.bucket, Key=self.key,
+                                         Range='bytes=%i-%i' % (start, self.start - 1)
+                                         )['Body'].read()
+                self.start = start
+                self.cache = new + self.cache
+            if end > self.end:
```

Comment:
You can request any block, including past the end of the file (like with BytesIO); that should return empty bytestrings and not error.